### PR TITLE
Tolerate overshooting during density tests

### PR DESCRIPTION
**Pull request by jayunit100**
This will allow us to proceed w/ testing catastrophic HA recovery scenarios (i.e. controller manager dies) without failing.   Otherwise, density can fail pretty easily in a large cluster where there might be pods transitioning to running, because it will overschedule (temporarily).

### CODE REVIEW

**Comment by jayunit100** [comment]
Allows us to survive pending complete fix to #10030

**Comment by rrati** [comment]
I would change 

fmt.Errorf("Only %d pods ...

and remove "Only" since you are now relying on this for overshoot as well.

File: test/e2e/util.go [hunk 1]
```diff
@@ -1156,6 +1160,10 @@ func RunRC(config RCConfig) error {
 			Diff(oldPods, pods).Print(util.NewStringSet())
 			return fmt.Errorf(errorStr)
 		}
+		//TODO: Convert this warning to failure after #10030 is fixed.
+		if len(pods) > config.Replicas {
+			LogfPrefix("WARN", "POD overshoot occured %d (actual pods) > %d (replicas) !!! See https://github.com/GoogleCloudPlatform/kubernetes/issues/10030. This is not necessarily a failure, but hopefully it will be repaired soon.", len(pods), config.Replicas)
+		}
```
Comment (rrati) [inline test/e2e/util.go]: you might want to add a condition here to break out if the system hasn't corrected itself in some time period.  You might be able to rely on the timeout to determine the system isn't going to correct itself.

File: test/e2e/util.go [hunk 2]
```diff
@@ -1145,7 +1149,7 @@ func RunRC(config RCConfig) error {
 			dumpNodeDebugInfo(config.Client, containerRestartNodes.List())
 			return fmt.Errorf("%d containers failed which is more than allowed %d", failedContainers, maxContainerFailures)
 		}
-		if len(pods) < len(oldPods) || len(pods) > config.Replicas {
+		if len(pods) < len(oldPods) {
 			// This failure mode includes:
```
Comment (rrati) [inline test/e2e/util.go]: Are both failure cases still relevant in this failure?  I think the first is (is accounts for the pod count drop), but the 2nd might be more applicable to the overshoot case.

Comment (jayunit100) [inline test/e2e/util.go]: Ah yes, well the plan is to merge this once we have overshooting solved... But yeah I guess clarifying the comment should be done

File: test/e2e/util.go (hunk 1 above)
Comment (jayunit100) [inline test/e2e/util.go]: but timeouts seem to mask the real bug,  after all don't we want no overshooting?  I can add timeout but just want to be sure we're adding for the right reasons

Comment (rrati) [inline test/e2e/util.go]: True, in the long run this would be an error and the test would fail but right now we are accepting that an overshoot can occur and expect that the system will recover.  What if the system doesn't recover?  The test might (depending on what the failure case looks like) continue to run forever with nothing to time it out.  Until we make this a hard failure I think we need to ensure the test won't run forever.

**Comment by bprashanth** [comment]
Can you do this only when HA=true? See  https://github.com/GoogleCloudPlatform/kubernetes/issues/10030#issuecomment-122041920 for details on overshoot scenario. We still want to fail on overshoot.

File: test/e2e/util.go [hunk 3]
```diff
@@ -186,6 +186,10 @@ type RCConfig struct {
 	MaxContainerFailures *int
 }
 
+func LogfPrefix(prefixx string, format string, a ...interface{}) {
```
Comment (wojtek-t) [inline test/e2e/util.go]: I don't understand the reason behind creating this function. Is it really needed?

File: test/e2e/util.go [hunk 4]
```diff
@@ -1156,6 +1160,10 @@ func RunRC(config RCConfig) error {
 			Diff(oldPods, pods).Print(util.NewStringSet())
 			return fmt.Errorf(errorStr)
 		}
+		//TODO: Convert this warning to failure after #10030 is fixed.
+		if len(pods) > config.Replicas {
+			LogfPrefix("WARN", "POD overshoot occured %d (actual pods) > %d (replicas) !!! See https://github.com/GoogleCloudPlatform/kubernetes/issues/10030. This is not necessarily a failure, but hopefully it will be repaired soon.", len(pods), config.Replicas)
```
Comment (wojtek-t) [inline test/e2e/util.go]: Why not just "Logf("WARNING: POD overshoot ...

File: test/e2e/util.go (hunk 3 above)
Comment (jayunit100) [inline test/e2e/util.go]: The other one does info by default

File: test/e2e/util.go (hunk 4 above)
Comment (jayunit100) [inline test/e2e/util.go]: We could but then it would be "INFO: WARNING..." From what I can tell because of the hardcoded info... Am I missing something?

File: test/e2e/util.go (hunk 1 above)
Comment (jayunit100) [inline test/e2e/util.go]: ack I'll put a timeout good idea

File: test/e2e/util.go (hunk 4 above)
Comment (wojtek-t) [inline test/e2e/util.go]: You're right, but still I think it's overcomplicating the code. Anyway, I'm not going to block the PR on this.

Comment (jayunit100) [inline test/e2e/util.go]: Ok. How about I remove INFO from Logf? Is that a better solution?

Comment (wojtek-t) [inline test/e2e/util.go]: SGTM

Comment (wojtek-t) [inline test/e2e/util.go]: Also (if it's easy) it would be great to have timestamps in the all the logs.

**Comment by jayunit100** [comment]
@bprashanth Is there a way we can inspect controller-manager API server or ETCD to determine if controller-manager was restarted ?  That way we dont need to specify "ha" as an external metadata, and we can tolerate overshoots in those cases .

**Comment by bprashanth** [comment]
Yeah, you can send an event at startup time and list events when you observe an overshoot (anywhere in this callstack https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L144). 

While that's useful, a controller manager restart is enough reason to fail any test that didn't explicitly kill it, overshoot or not. Since this is density and you are manually killing it to trigger HA (right?), I'd rather not ignore overshoots by default (if you don't want to add a --ignore flag I'll take another addition to https://github.com/GoogleCloudPlatform/kubernetes/blob/master/test/e2e/density.go#L168 for now).

Basically I don't want our e2e runs papering over controller manager restarts and/or replica overshoots.

**Comment by timothysc** [comment]
@jayunit100 @bprashanth - perhaps we should re-purpose and automate a failover case instead and copy density as part of it?  I'm guessing the HA set of tests will grow...

Thoughts?

**Comment by jayunit100** [comment]
Yup - thats a good way to to id .  ROOT CAUSE = util.go deciding wether or not a test fails or succeeds - the coupling is causing a problem.   So lets let RunRC return some info to the calling test, and then using @timothysc 's idea,  HA density tests can have more lax failure criteria for now.

**Comment by timothysc** [comment]
@jayunit100 ideally we can trip a failover on the test.

**Comment by jayunit100** [comment]
how? thats a great idea.

**Comment by bprashanth** [comment]
Fyi if you want something that can restart daemons easily you might just wait for https://github.com/GoogleCloudPlatform/kubernetes/pull/11339

**Comment by jayunit100** [comment]
summarizing: 

- bprashant: we need the test to fail on overshoots
- bprashant: but we want the test to allow overshoots if doing HA, because that is a known consequence of HA right now.
- jayunit100: util.go is failing tests , embedding logic on when to fail, **needs decoupling** for pure reasons, and also benefits this particular issue.
- timothysc: we should just have a go test which knows its going to overshoot and parameterizes runRC accordingly
- bprashant: there is a PR in flight #11339 which can trip failover, which means we can consolidate failover test into one go method.  This allows us to have a smart ha density test which trips itself.  Even until then, decoupled RC can simply have a ginkgo filter so we can run HA density tests .

all in agreement?

**Comment by bprashanth** [comment]
SGTM. I'm happy as long as the common case (our hourly jenkins runs) don't gloss over bugs :)

The proposed solution of adding an HA ginkgo flag and tripping e2e tests (doesn't have to be just density right) via daemon restart triggered by some restart-y framework, makes sense as well.

**Comment by timothysc** [comment]
+1. 

> I'm happy as long as the common case (our hourly jenkins runs) don't gloss over bugs :) 
Never happens ;-)

**Comment by jayunit100** [comment]
great ! ill update this test accordingly to have an HA Variant (makes tim happy), and parameterize RunRC (makes me happy).  And the net result will be a density test which is strict (makes bprashant happy).  SO MUCH HAPPINESS.

**Comment by jayunit100** [comment]
rob brought up the fact that Error isn't ACTUALLY causing test fail, so... maybe this is a no-op and we should close - and just interpret HA test results with a grain of salt knowing that they will overshoot.?

**Comment by jayunit100** [comment]
CLOSED ! This really is a no-op.  We need Density to continue failing when overshoots happen, and the simplest way to deal with this is by interpretting the test results proactively when doing ha tests.
